### PR TITLE
Adding testng variant, supporting through IExecutionListener

### PR DESCRIPTION
**Pull request by santthosh**
## Description
@agudian Adding TestNG variant

## Tasks
- [x] I have signed the [CLA](http://adobe.github.io/cla.html).
- [x] I have written tests and verified that they fail without my change.

### CODE REVIEW

File: testsupport/testng/pom.xml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                    <artifactId>maven-surefire-plugin</artifactId>
+                    <!-- needs to be pinned down to 2.19.1 until a new junit-platform-surefire-provider provider is released -->
+                    <version>2.19.1</version>
+                    <dependencies>
+                        <dependency>
+                            <groupId>org.apache.maven.surefire</groupId>
+                            <artifactId>surefire-testng</artifactId>
+                            <version>2.19.1</version>
+                        </dependency>
+                    </dependencies>
```
Comment (agudian) [inline testsupport/testng/pom.xml]: You can safely remove the comment, the pinning to version 2.19.1 and the `<dependencies>` section here. The automatic discovery of the testng surefire provider works correctly, as no JUnit dependency is on the classpath here.

File: testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+public class S3Mock extends S3MockStarter {
+    private static S3Mock instance = null;
+
+    private S3Mock() {
+        super(null);
+        this.properties.putAll(defaultProps());
+        final Map<String, Object> properties = new HashMap<>();
+
+        String httpsPort =  System.getProperty(S3MockApplication.PROP_HTTPS_PORT);
```
Comment (agudian) [inline testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java]: Isn't that working already? If not, we should fix that directly in `S3MockApplication`. After that, you wouldn't need to do anything else but call `super(null)` in this constructor.

File: testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3MockListener.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+package com.adobe.testing.s3mock.testng;
+
+import org.testng.IExecutionListener;
+
+/**
+ * TestNG listener to start and stop the S3Mock Application. After the tests, the S3Mock is
+ * stopped.
+ * <p>
```
Comment (agudian) [inline testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3MockListener.java]: please remove the `<p>`, it's superfluous here.

File: testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class S3Mock extends S3MockStarter {
+    private static S3Mock instance = null;
+
+    private S3Mock() {
+        super(null);
+        this.properties.putAll(defaultProps());
```
Comment (agudian) [inline testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java]: `defaultProps()` have already been applied.

File: testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        String rootDirectory = System.getProperty(S3MockApplication.PROP_ROOT_DIRECTORY);
+        if(rootDirectory != null) {
+            properties.put(S3MockApplication.PROP_ROOT_DIRECTORY,rootDirectory);
+        }
+
+        this.properties.putAll(properties);
+    }
+
+    public static S3Mock getInstance() {
+        if(instance == null) instance = new S3Mock();
```
Comment (agudian) [inline testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java]: To make it thread-safe, you could initialize `instance` directly when declaring it.

File: testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+package com.adobe.testing.s3mock.testng;
+
+import com.adobe.testing.s3mock.S3MockApplication;
+import com.adobe.testing.s3mock.testsupport.common.S3MockStarter;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class S3Mock extends S3MockStarter {
```
Comment (agudian) [inline testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java]: Please add some Javadoc, linking to `S3MockListener` as a requirement to use this class and maybe mentioning that it used as a singleton.

**Comment by santthosh** [comment]
Thanks @agudian, can you please review again?

File: testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    private static S3Mock instance = new S3Mock();
+
+    private S3Mock() {
+        super(null);
+    }
+
+    public static S3Mock getInstance() {
+        if(instance == null) instance = new S3Mock();
```
Comment (agudian) [inline testsupport/testng/src/main/java/com/adobe/testing/s3mock/testng/S3Mock.java]: This check can be removed and `instance` can be made final, right? And renamed to INSTANCE then?

**Comment by santthosh** [comment]
Thanks @agudian, did that too, sorry have been more close to javascript, still catching up with best practices. Let me know if I can be of more help

**Pull request merged**